src/GNDStk/Node.hpp

```diff
@@ -0,0 +1,187 @@
+class TypedNode;
+
+
+
+// -----------------------------------------------------------------------------
+// Node (templated)
+// node = Node<>
+// -----------------------------------------------------------------------------
+
+template<
+   template<class...> class METADATA_CONTAINER, // container type for metadata
+   template<class...> class CHILDREN_CONTAINER  // container type for children
+>
+class Node {
+   using metaPair = std::pair<std::string,std::string>;
+   using childPtr = std::unique_ptr<Node>;
+
+public:
+
+   // ------------------------
+   // Data
+   // ------------------------
+
+   // Simple node for our tree structure:
+   //    name
+   //    metadata (container of string pairs)
+   //    children (container of pointers to other Node<>s)
+   std::string name;
+   METADATA_CONTAINER<metaPair,std::allocator<metaPair>> metadata;
+   CHILDREN_CONTAINER<childPtr,std::allocator<childPtr>> children;
+
+   // ------------------------
+   // Simple functions
+   // ------------------------
+
+   // clear()
+   Node &clear()
+   {
+      name = "";
+      metadata.clear();
+      children.clear();
+      return *this;
+   }
+
+   // is this node empty?
+   bool empty() const
+   {
+      return name == "" && metadata.size() == 0 && children.size() == 0;
+   }
+
+   // is this a leaf node?
+   bool leaf() const
+   {
+      return children.size() == 0;
+   }
+
+   // copy
+   Node copy() const
+   {
+      Node ret;
+      detail::node2Node(*this,ret);
+      return ret;
+   }
+
+   // normalize
+   Node &normalize();
+
+   // ------------------------
+   // General functions
+   // ------------------------
+
+   // ctor: default
+   Node() { }
+
+   // ctor: move, DELETED copy
+   Node(Node &&) = default;
+   Node(const Node &) = delete;
+
+   // assignment: move, DELETED copy
+   Node &operator=(Node &&) = default;
+   Node &operator=(const Node &) = delete;
+
+   #include "GNDStk/Node/src/add.hpp"
+   #include "GNDStk/Node/src/write.hpp"
+   #include "GNDStk/Node/src/meta.hpp"
+   #include "GNDStk/Node/src/child.hpp"
+   #include "GNDStk/Node/src/call.hpp"
+
+}; // class Node
+
+
+
+// -----------------------------------------------------------------------------
+// operator<<
+// -----------------------------------------------------------------------------
+
+template<
+   template<class...> class METADATA_CONTAINER,
+   template<class...> class CHILDREN_CONTAINER
+>
+inline std::ostream &operator<<(
+   std::ostream &os,
+   const Node<METADATA_CONTAINER,CHILDREN_CONTAINER> &obj
+) {
+   return obj.write(os);
+}
+
+
+
+// -----------------------------------------------------------------------------
+// normalize
+// -----------------------------------------------------------------------------
+
+template<
+   template<class...> class METADATA_CONTAINER,
+   template<class...> class CHILDREN_CONTAINER
+>
+Node<METADATA_CONTAINER,CHILDREN_CONTAINER> &
+Node<METADATA_CONTAINER,CHILDREN_CONTAINER>::normalize()
```

Comment:
Probably two files, as this needs to be split up (and renamed) as well. I plan to get to the JSON business soon. At present, normalize() is about dealing with the fact that JSON files shoehorn metadata into child nodes, because JSON doesn't support attributes directly. normalize() un-shoehorns the information. What I'll really need is something like an attributes2children() plus also a children2attributes(). That's been on my to-do list, underneath more-pressing things.

Reply:
Revisiting JSON is the next major thing on the to-do list. Wanted to tidy up some other things first.

Reply:
Or maybe Node tests. Both seem similarly important.